File: services/settlement/internal/store/oz_ledger.go
// Go-side Oz ledger writer. Mirrors the TS primitive in
// services/api/src/modules/community/oz-ledger.ts — the two MUST stay
// SQL-compatible because both can write the same table (TS via the
// admin credit endpoint and future engagement-floor hook; Go via the
// settler tx). The shared idempotency_key UNIQUE constraint makes the
// two-writer arrangement safe: same key = exactly one row, regardless
// of which writer got there first.
//
// V1 mode is earn-only — the DB CHECK constraint enforces delta > 0
// (migration 0076). This file does not need to handle spend; when the
// redemption flow lands it will likely live entirely on the TS side
// (admin redemption / user-initiated spend), and this Go writer stays
// credit-only.
//
// Why a Go-side writer (instead of emitting a settlement event for a
// TS subscriber to consume): keeps the credit atomic with the
// settlement transaction. A settle that succeeds but whose post-hoc
// Oz credit fails would either lose the credit or require a second
// reconciliation worker. Writing inline removes that class of
// failure.

package store

import (
	"context"
	"fmt"

	"github.com/jackc/pgx/v5"
)

// CreditOzInput mirrors the TS CreditOzInput type. Note `Delta` is an
// int64 (not bigint) because realistic per-credit values are far below
// 2^63 — the Reward formula V1 cap is 250 Oz per credit. The column
// is BIGINT for header-room only.
type CreditOzInput struct {
	UserID          string
	Delta           int64
	Reason          string
	SourceKind      string // "analysis" | "admin" | "" if absent
	SourceID        string // empty if absent
	IdempotencyKey  string
	CreatedBy       string // empty if absent (system credits)
}

// CreditOzResult mirrors the TS shape. Credited=false on idempotency-
// key collision; never an error.
type CreditOzResult struct {
	Credited     bool
	BalanceAfter int64
}

// CreditOz inserts a ledger row and upserts oz_balance_user atomically.
// Same CTE shape as the TS primitive — the SQL is intentionally copy-
// paste-compatible so future audit tools can diff the two writers.
//
// Caller MUST construct a deterministic idempotency key from the
// (reason, source) pair. A second call with the same key returns
// `{Credited: false, BalanceAfter: <current balance>}` rather than an
// error.
func CreditOz(ctx context.Context, tx pgx.Tx, input CreditOzInput) (CreditOzResult, error) {
	if input.Delta <= 0 {
		return CreditOzResult{}, fmt.Errorf("oz_credit_invalid_delta: %d", input.Delta)
	}

	// nilIfEmpty bridges Go's "" semantics to SQL NULL — we want
	// source_kind / source_id / created_by NULL when not set rather
	// than empty strings.
	var sourceKind, sourceID, createdBy any
	if input.SourceKind != "" {
		sourceKind = input.SourceKind
	}
	if input.SourceID != "" {
		sourceID = input.SourceID
	}
	if input.CreatedBy != "" {
		createdBy = input.CreatedBy
	}

	const q = `
WITH ins AS (
  INSERT INTO oz_ledger
    (user_id, delta, reason, source_kind, source_id, idempotency_key, created_by)
  VALUES
    ($1::uuid, $2::bigint, $3::text, $4::text, $5::text, $6::text, $7::uuid)
  ON CONFLICT (idempotency_key) DO NOTHING
  RETURNING delta
),
bump AS (
  INSERT INTO oz_balance_user (user_id, balance, updated_at)
  SELECT $1::uuid, ins.delta, NOW()
    FROM ins
  ON CONFLICT (user_id) DO UPDATE
    SET balance    = oz_balance_user.balance + EXCLUDED.balance,
        updated_at = NOW()
  RETURNING balance
)
SELECT
  EXISTS (SELECT 1 FROM ins)                                AS credited,
  COALESCE(
    (SELECT balance FROM bump),
    (SELECT balance FROM oz_balance_user WHERE user_id = $1::uuid),
    0
  )::bigint                                                 AS balance_after
`
	var res CreditOzResult
	if err := tx.QueryRow(
		ctx, q,
		input.UserID,
		input.Delta,
		input.Reason,
		sourceKind,
		sourceID,
		input.IdempotencyKey,
		createdBy,
	).Scan(&res.Credited, &res.BalanceAfter); err != nil {
		return CreditOzResult{}, fmt.Errorf("oz credit: %w", err)
	}
	return res, nil
}

// OzFromStakeMicros mirrors the TS helper in analyses-rewards.ts.
// Stake-scaled Oz reward: 25% of stake in stake units (micros / 1M),
// floored at 1 Oz, capped at 250. Currency-agnostic by construction.
//
// Used by both the engagement-floor credit (TS) and the win-bonus
// credit (Go). Two callers, one formula — kept in sync by code
// review across the language boundary.
func OzFromStakeMicros(stakeMicro int64) int64 {
	stakeUnits := stakeMicro / 1_000_000
	raw := stakeUnits / 4 // * 0.25
	if raw < 1 {
		return 1
	}
	if raw > 250 {
		return 250
	}
	return raw
}

File: services/api/src/modules/community/oz-ledger.ts
// Oz loyalty-point ledger primitive.
//
// Every Oz credit in the system flows through `creditOz`. Callers
// construct a deterministic `idempotencyKey` from the (reason, source)
// pair and rely on the UNIQUE constraint on `oz_ledger.idempotency_key`
// (migration 0076) to make replay safe. On collision the function
// returns `{ credited: false }` — never throws — so settlement retries
// and admin double-clicks both no-op cleanly.
//
// The ledger row + balance upsert run inside whichever transaction the
// caller passes. Mirroring the writeCommunityProjection convention so
// settlement-side hooks (Go-driven, when they land) and TS-side hooks
// (engagement-floor, admin endpoint) can both participate in the
// triggering transaction without a separate connection.

import { sql } from "drizzle-orm";
import type { DbClient } from "@oddzilla/db";

// Same structural type as writeCommunityProjection — accepts both the
// request-scoped client and a transaction handle.
export type OzLedgerExecutor = Pick<DbClient, "execute">;

export interface CreditOzInput {
  userId: string;
  delta: number;
  reason: string;
  sourceKind?: string | null;
  sourceId?: string | null;
  idempotencyKey: string;
  // For admin credits, the admin user id. NULL on system credits
  // (engagement-floor / win-bonus / future automated hooks).
  createdBy?: string | null;
}

export interface CreditOzResult {
  // false when the idempotency key already exists — the caller's
  // retry hit a row that was previously credited. Not an error.
  credited: boolean;
  // Current balance after the credit (or the pre-existing balance on
  // a dedup'd call). Read out of oz_balance_user in the same round
  // trip so callers don't need a follow-up SELECT.
  balanceAfter: number;
}

// One round-trip implementation. The CTE chain is:
//   1. ins  — INSERT … ON CONFLICT (idempotency_key) DO NOTHING
//             RETURNING delta. Empty on dedup hit.
//   2. bump — INSERT into oz_balance_user with delta on conflict
//             update balance += delta. Skipped when ins is empty
//             via the WHERE EXISTS gate.
//   3. SELECT current balance for the user.
//
// Doing it as one statement keeps the credit atomic against parallel
// credits to the same user: a second concurrent call sees the bumped
// balance, not a phantom pre-update read.
export async function creditOz(
  db: OzLedgerExecutor,
  input: CreditOzInput,
): Promise<CreditOzResult> {
  if (!Number.isInteger(input.delta) || input.delta <= 0) {
    // Defence-in-depth — the DB CHECK constraint also enforces this,
    // but catching at the API boundary surfaces the bug at the
    // caller's stack frame, not inside Postgres.
    throw new Error(`oz_credit_invalid_delta: ${input.delta}`);
  }

  const rows = await db.execute<{ credited: boolean; balance_after: number }>(sql`
WITH ins AS (
  INSERT INTO oz_ledger
    (user_id, delta, reason, source_kind, source_id, idempotency_key, created_by)
  VALUES
    (${input.userId}::uuid,
     ${input.delta}::bigint,
     ${input.reason}::text,
     ${input.sourceKind ?? null}::text,
     ${input.sourceId ?? null}::text,
     ${input.idempotencyKey}::text,
     ${input.createdBy ?? null}::uuid)
  ON CONFLICT (idempotency_key) DO NOTHING
  RETURNING delta
),
bump AS (
  INSERT INTO oz_balance_user (user_id, balance, updated_at)
  SELECT ${input.userId}::uuid, ins.delta, now()
    FROM ins
  ON CONFLICT (user_id) DO UPDATE
    SET balance    = oz_balance_user.balance + EXCLUDED.balance,
        updated_at = now()
  RETURNING balance
)
SELECT
  EXISTS (SELECT 1 FROM ins)                                AS "credited",
  COALESCE(
    (SELECT balance FROM bump),
    (SELECT balance FROM oz_balance_user WHERE user_id = ${input.userId}::uuid),
    0
  )::bigint                                                 AS "balance_after"
  `);

  const row = rows[0];
  if (!row) {
    // Shouldn't happen — the SELECT always returns one row — but
    // surface as an error rather than silently returning a stub.
    throw new Error("oz_credit_no_result");
  }
  return {
    credited: Boolean(row.credited),
    balanceAfter: Number(row.balance_after),
  };
}
